src/Dscp/DB/SQLite/Queries.hs

```diff
@@ -140,6 +149,134 @@ getStudentTransactions student = do
+getProvenStudentTransactionsSince studentId sinceTime = do
+    transaction $ do
+        -- Contains `(tx, idx, blockId)` map.
+        txsBlockList <- getTxsBlockMap
+
+        -- Bake `blockId -> [(tx, idx)]` map.
+        let txsBlockMap = groupToAssocWith (_tibBlockId, _tibTx) txsBlockList
+
+        forM txsBlockMap $ \(blockId, transactions) -> do
+            blockData <- getBlockData blockId
+
+            let indices = _twiTxIdx <$> transactions & Set.fromList
+            let bodies  = _twiTx    <$> transactions
+            let tree    = _bdTree blockData
+            let pruned  = mkMerkleProof tree indices
+
+            return (pruned, bodies)
+
+  where
+    groupToAssocWith :: Eq k => Ord k => (a -> k, a -> v) -> [a] -> [(k, [v])]
+    groupToAssocWith (key, value)
+        -- The `groupBy` should return `[NonEmpty a]`.
+        = map      (\every@ (selected : _) -> (key selected, map value every))
+        . groupBy  ((==) `on` key)
+        . sortWith  key
+
+    -- Returns, effectively, `[((PrivateTx, Idx), Id PrivateBlock)]`.
+    getTxsBlockMap :: DBM m => m [TxInBlock]
+    getTxsBlockMap = query getTxsBlockMapQuery (studentId, sinceTime)
+      where
+        getTxsBlockMapQuery = [q|
+            -- getTxsBlockMapQuery
+
+            select     Submissions.student_addr,
+                       Submissions.contents_hash,
+                       Assignments.course_id,
+                       Assignments.contents_hash,
+                       Assignments.type,
+                       Assignments.desc,
+                       Submissions.signature,
+                       Transactions.grade,
+                       Transactions.time,
+                       Transactions.idx
+                       BlockTxs.blk_idx
+
+            from       BlockTxs
+            left join  Transactions
+                   on  tx_hash = Transactions.hash
+
+            left join  Submissions
+                   on  Transactions.submission_hash = Submissions.hash
+
+            left join  Assignments
+                   on  Submissions .assignment_hash = Assignments.hash
+
+            where      student_addr  = ?
+                  and  time         >= ?
+                  and  idx          <> -1
+        |]
+
+    -- Returns `PrivateBlock` in normalized format, with metadata.
+    getBlockData :: DBM m => Id PrivateBlock -> m BlockData
+    getBlockData blockId = do
+        (listToMaybe <$> query getBlockDataQuery (Only blockId))
+            `assertJust` BlockDoesNotExist blockId
+      where
+        -- TODO: prune fields that aren't needed.
+        getBlockDataQuery = [q|
+            select  idx
+                    hash
+                    time
+                    prev_hash
+                    atg_delta
+                    mroot
+                    mtree
+            from    Blocks
+            where   idx = ?
+        |]
+
+createBlock :: DBM m => PrivateBlock -> m (Id PrivateBlock)
+createBlock block = do
```

Comment:
The block will always be constructed from the set of all currently blockless transactions (those which don't have a corresponding record in `BlockTxs` table). Why then not have the type `createBlock :: DBM m => m PrivateBlock` and perform fetching these them and constructing the block in the same DB transaction with saving it back to the DB?

Reply:
I left one `Maybe ATGDelta` param there.